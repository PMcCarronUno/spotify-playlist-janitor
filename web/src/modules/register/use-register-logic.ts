import { useAuthApi } from "api/auth-api";
import { useCallback, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LOGIN } from "shared/constants";

export const useRegisterLogic = () => {
  const [error, setError] = useState<string | undefined>();
  const [disabled, setDisabled] = useState(false);

  const navigate = useNavigate();
  const { register } = useAuthApi();

  const onSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      const email = (event.currentTarget[0] as HTMLInputElement).value;
      const spotifyEmail = (event.currentTarget[1] as HTMLInputElement).value;
      const password = (event.currentTarget[2] as HTMLInputElement).value;
      const confirmPassword = (event.currentTarget[3] as HTMLInputElement)
        .value;

      if (!email) {
        setError("Please enter email");
        return;
      }

      if (!password) {
        setError("Please enter password");
        return;
      }

      if (!confirmPassword) {
        setError("Please confirm password");
        return;
      }

      if (password !== confirmPassword) {
        setError("Passwords do not match");
        return;
      }

      try {
        setDisabled(true);
        await register({
          email,
          spotifyEmail: !!spotifyEmail ? spotifyEmail : email,
          password,
        });

        setError(undefined);
        navigate(LOGIN);
      } catch (e: any) {
        setError(e.response?.data?.message || "Unknown error");
      } finally {
        setDisabled(false);
      }
    },
    [setError, register, navigate]
  );

  return { disabled, onSubmit, error } as const;
};
