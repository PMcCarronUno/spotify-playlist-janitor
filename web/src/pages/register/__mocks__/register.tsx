import React from 'react';
export const Login = () => <div>Register Page</div>;