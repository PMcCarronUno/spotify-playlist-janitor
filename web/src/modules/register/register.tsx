import { PrimaryButton } from "shared/components/button";
import { TextInput } from "shared/components/text-input";
import { SecondaryText, Text, Title } from "shared/components/typography";
import { RED, WHITE } from "shared/constants";
import styled from "styled-components";
import { useRegisterLogic } from "./use-register-logic";

export const Register = () => {
  const { disabled, onSubmit, error } = useRegisterLogic();
  return (
    <PageContainer>
      <div className="row">
        <Title>Register</Title>
      </div>
      <div className="row">
        <Text>Register with email/password:</Text>
      </div>
      <form {...{ onSubmit }} autoComplete="off">
        <div className="row">
          <TextInput
            {...{
              name: "email",
              label: "Email",
              placeholder: "Enter email...",
              id: "email-input",
              "data-testid": "email-input",
              disabled,
            }}
          />
        </div>
        <div className="row">
          <TextInput
            {...{
              name: "password",
              type: "password",
              label: "Password",
              placeholder: "Enter password...",
              id: "password-input",
              "data-testid": "password-input",
              disabled,
            }}
          />
        </div>
        <div className="row">
          <TextInput
            {...{
              name: "passwordConfirm",
              type: "password",
              label: "Confirm Password",
              placeholder: "Enter password...",
              id: "password-confirm-input",
              "data-testid": "password-confirm-input",
              disabled,
            }}
          />
        </div>
        {error && (
          <div
            {...{
              className: "row error",
              id: "submit-error",
              "data-testid": "submit-error",
            }}
          >
            <SecondaryText>{error}</SecondaryText>
          </div>
        )}
        <div className="row">
          <PrimaryButton
            {...{
              id: "register-button",
              "data-testid": "register-button",
              disabled,
            }}
          >
            Register
          </PrimaryButton>
        </div>
      </form>
    </PageContainer>
  );
};

const PageContainer = styled.div`
  display: flex;
  flex-direction: column;
  color: ${WHITE};
  justify-content: center;
  align-items: center;

  form {
    width: 400px;
  }

  .row {
    margin-bottom: 8px;
    display: flex;
    justify-content: center;

    div {
      width: 260px;
      display: flex;
    }
  }

  .error {
    color: ${RED};
  }
`;

Register.displayName = "Register";
