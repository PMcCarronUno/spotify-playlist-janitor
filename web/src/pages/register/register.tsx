import { HEADER_HEIGHT, HEADER_PADDING } from "shared/constants";
import styled from "styled-components";
import { Register as RegisterInitialise } from "modules/register";

export const Register = () => {
  return (
    <PageContainer>
      <RegisterInitialise />
    </PageContainer>
  );
};

const PageContainer = styled.div`
  display: flex;
  justify-content: center;
  height: calc(
    100vh - ${HEADER_HEIGHT} - ${HEADER_PADDING} - ${HEADER_PADDING}
  );
  width: 400px;
  margin: auto;
`;

Register.displayName = "Register";
