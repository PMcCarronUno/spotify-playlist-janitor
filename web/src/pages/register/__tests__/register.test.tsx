import { render } from '@testing-library/react';

import { Register } from '../register';

jest.mock("modules/register");

describe('<Register />', () => {
  let container: HTMLElement;

  beforeEach(() => {
    ({ container } = render(<Register />));
  });

  it('should render a Register component', () => {
    expect(container.firstChild).toMatchSnapshot();
  });
});