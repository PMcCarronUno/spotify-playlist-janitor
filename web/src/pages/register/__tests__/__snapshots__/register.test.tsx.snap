// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> should render a Register component 1`] = `
<div
  class="sc-beyTiQ gLUzuQ"
>
  <div>
    Register Component
  </div>
</div>
`;
